packages/react-router/src/scroll-restoration.tsx

```diff
@@ -87,110 +190,84 @@ export function useScrollRestoration(options?: ScrollRestorationOptions) {
-              ...c,
-              next,
-              cached: {
-                ...c.cached,
-                [[restoreKey, elementSelector].join(delimiter)]: entry,
-              },
-            }
-          })
-        }
+    const unsubOnResolved = router.subscribe('onRendered', (event) => {
+      if (!router.resetNextScroll) {
+        return
       }
-    })
 
-    const unsubOnBeforeRouteMount = router.subscribe(
-      'onBeforeRouteMount',
-      (event) => {
-        if (event.hrefChanged) {
-          if (!router.resetNextScroll) {
-            return
-          }
+      router.resetNextScroll = true
 
-          router.resetNextScroll = true
-
-          const restoreKey = getKey(event.toLocation)
-          let windowRestored = false
-
-          for (const cacheKey in cache.state.cached) {
-            const entry = cache.state.cached[cacheKey]!
-            const [key, elementSelector] = cacheKey.split(delimiter)
-            if (key === restoreKey) {
-              if (elementSelector === windowKey) {
-                windowRestored = true
-                window.scrollTo({
-                  top: entry.scrollY,
-                  left: entry.scrollX,
-                  behavior: options?.scrollBehavior,
-                })
-              } else if (elementSelector) {
-                const element = document.querySelector(elementSelector)
-                if (element) {
-                  element.scrollLeft = entry.scrollX
-                  element.scrollTop = entry.scrollY
-                }
-              }
-            }
-          }
-
-          if (!windowRestored) {
-            window.scrollTo(0, 0)
-          }
+      restoreScroll(
+        storageKey,
+        getKey(event.toLocation),
+        options?.scrollBehavior,
+      )
+    })
 
-          cache.set((c) => ({ ...c, next: {} }))
-          weakScrolledElements = new WeakSet<any>()
-        }
-      },
+    restoreScroll(
+      storageKey,
+      getKey(router.latestLocation),
+      options?.scrollBehavior,
     )
 
     return () => {
-      document.removeEventListener('scroll', onScroll)
-      unsubOnBeforeLoad()
-      unsubOnBeforeRouteMount()
+      unobserveDom()
+      document.removeEventListener('scroll', throttle(onScroll, 100))
+      unsubOnResolved()
     }
   }, [options?.getKey, router])
 }
 
 export function ScrollRestoration(props: ScrollRestorationOptions) {
+  const router = useRouter()
   useScrollRestoration(props)
-  return null
+
+  // if (!router.) {
+  //   return <script suppressHydrationWarning />
+  // }
+
+  const getKey = props.getKey || defaultGetKey
+  const userKey = getKey(router.latestLocation)
+  const resolvedKey =
+    userKey !== defaultGetKey(router.latestLocation) ? userKey : null
+
+  return (
+    <script
+      suppressHydrationWarning
+      dangerouslySetInnerHTML={{
+        __html: `(${restoreScroll.toString()})(${JSON.stringify(storageKey)},${JSON.stringify(resolvedKey)});__TSR__.cleanScripts()`,
```

Comment:
Perhaps this bit should be optionally checked since Start isn't guaranteed to be there?
```suggestion
        __html: `(${restoreScroll.toString()})(${JSON.stringify(storageKey)},${JSON.stringify(resolvedKey)});__TSR__?.cleanScripts?.()`,
```

Reply:
it's also called `__TSR_SSR__` now